### DISCUSSION TITLE
transfer-06-consumer-pull-http: questions about EndpointDataReference

### DISCUSSION THREAD

**Discussion by johnnychoque** [eclipse-edc/Samples#113] [Q&A]
Hi all,

While running the steps of the transfer-06-consumer-pull-http sample, I noticed that the http server logs show the following EndpointDataReference:

```
{
  "id": "5b3209bc-f36b-43f1-ab24-7fa83a805180",
  "endpoint": "http://localhost:19291/public/",
  "authKey": "Authorization",
  "authCode": 
     "eyJhbGciOiJSUzI1NiJ9.eyJleHAiOjE2OTQxOTE2MTcsImRhZCI6IntcInByb3BlcnRpZXNcIjp7
      XCJodHRwczovL3czaWQnL2VkYy92MC4wLjEcHJveHlQYXRoXCI6XCJ0c. . . k2QIwoWXNz_lRg",
  "properties": {}
}
```

In that json the endpoint property has port 19291, while the documentation of the sample (step 11) shows port 29291. I have used the endpoint with those ports to get the data from the provider and it works with both. Wouldn't it make more sense for the consumer can get the data from the provider through an endpoint with a provider port (19291)?

Trying to achieve the above, I have removed the publicApiUrl parameter from the consumer data plane [https://github.com/eclipse-edc/Samples/blob/main/transfer/transfer-06-consumer-pull-http/README.md#2-register-data-plane-instance-for-consumer](url). 

```
curl -H 'Content-Type: application/json' \
     -d '{
           "@context": {
             "edc": "https://w3id.org/edc/v0.0.1/ns/"
           },
           "@id": "http-pull-consumer-dataplane",
           "url": "http://localhost:29192/control/transfer",
           "allowedSourceTypes": [ "HttpData" ],
           "allowedDestTypes": [ "HttpProxy", "HttpData" ],
           "properties": {

            }
         }' \
             -X POST "http://localhost:29193/management/v2/dataplanes"
```

But when I run the curl command in step 11 I still get the data from the provider using the endpoint with the consumer port (29291). Why doesn't removing the publicApiUrl parameter from the consumer data plane have any effect?

**Comment by oliverbusch** [comment]
@johnnychoque  That's not a direct answer to your question, but it may lead us to an answer.

I tested the sample with the Tractus-X MXD and was surprised that it did not work 1:1 there. 

The call to the consumer data plane
`curl --location --request GET 'http://localhost:29291/public/' --header 'Authorization: <auth code>'`
leads to an error. 

The call to the provider data plane
`curl --location --request GET 'http://localhost:19291/public/' --header 'Authorization: <auth code>'`
returns a result here. 

My first understanding based on the [example](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-02-consumer-pull) was that the consumer backend service never communicates directly with the provider data plane, and that communication always goes through the consumer data plane. This is how it looks in the EDC example [transfer-02-consumer-pull](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-02-consumer-pull)  .

`curl --location --request GET 'http://localhost:29291/public/' --header 'Authorization: <auth code>'`

However, it looks different in the transfer-data-plane-consumer-pull [sequence diagram](https://github.com/eclipse-edc/Connector/blob/main/docs/developer/diagrams/transfer-data-plane-consumer-pull.png).

![303573618-20934bc3-c886-492f-898f-113b3694f077](https://github.com/eclipse-edc/Samples/assets/12880317/80dd7116-379a-42ee-ac8d-cd1745da395c)

According to the sequence diagram, the call would work according to your description. Here the consumer can get the data from the provider through an endpoint with provider port (19291)?

This begs the following questions:
1. Is the description in the sample wrong?
2. Why does the communication not work according to the sequence diagram?

Note: The data planes are already configured in the Tractus-X MXD. So, this can be a reason, that the call to the consumer data plane does not work.
